### ISSUE TITLE
Tracking upstream issue QTBUG-85498

### ISSUE THREAD

**Issue by sssoleileraaa** [freedomofpress/securedrop-client#815]
# Description

This is an issue to track upstream issue: https://bugreports.qt.io/browse/QTBUG-85498

We use `QLabel`s in various places, including our message and reply speech bubbles. `QLabel` provides a wrapping feature but does not automatically wrap strings that are longer than the width of the label and contains no spaces, dashes, and other unknown wrapping-point characters (wrapping points are not documented or specified in the Qt code so it's a bit of an unknown).

# More info

QLabel knows to wrap based on the length of a line of text and where the nearest wrapping-point character is within the bounds of the width of the label. Unfortunately, there isn't Qt support to adjust the wrapping policy to say that we prefer wrapping over cutting off strings that don't contain wrapping points.

We've learned that `QTextEdit` and `QPlainTextEdit` support the kind of wrapping we are interested in by default, however as you can see in #1050 there is the problem with recalculating and setting the height of a `QTextEdit` to fit the contents of its internal document.

Test data can be found here: https://github.com/freedomofpress/securedrop-client/wiki/Test-plan#input-testing

**Comment by creviera** [comment]
I think the desired behavior is to use word wrapping over a horizontal scrollbar

**Comment by eloquence** [comment]
Marked as release blocker due to information loss from the user's perspective.

**Comment by eloquence** [comment]
Pulled into current sprint per discussion on #799.

**Comment by ntoll** [comment]
I've been able to reproduce. I suspect a word-wrap strategy flag needs to be set to the appropriate value. Will try to get as close to @creviera's suggestion as possible in coming up PR. :+1:

**Comment by ntoll** [comment]
Hmmm... https://doc.qt.io/qt-5/qlabel.html#wordWrap-prop :-(

It looks like Qt's default word-wrap only works on whitespace, not as @creviera specifies. However, I notice (see screenie) that if I `setWordWrap(True)` on the QLabel in our `SpeechBubble` widget I notice that things do start to wrap, giving rise to the possibility that if there is a single word > X characters (where X is our current max char width) then we insert a "-" at that point so Qt's word-wrapping kicks in. I'll code up this approach into a PR.

**Comment by ntoll** [comment]
FWIW, this also effects summary and messages too:

![messagewrap](https://user-images.githubusercontent.com/37602/75676036-57649400-5c80-11ea-84e4-c133f4f34a26.png)

**Comment by ntoll** [comment]
Thank you St.Guido van Rossum: https://docs.python.org/3.8/library/textwrap.html

**Comment by eloquence** [comment]
Reopening per findings in #998.

**Comment by creviera** [comment]
QLabel knows to wrap based on the length of a line of text and where the nearest wrapping-point character is within the bounds of the width of the label. Unfortunately, there isn't Qt support to adjust the wrapping policy to say that we prefer wrapping over cutting off strings that don't contain wrapping points. 

We've learned that `QTextEdit` and `QPlainTextEdit` support the kind of wrapping we are interested in by default, however as you can see in #1050 there is the problem with recalculating and setting the height of a `QTextEdit` to fit the contents of its internal document.

As of now, we've looked into the following possible solutions:

* Use our own wrapping method that uses the library called `textwrap` to wrap message text every 70 characters. We saw problems with the way this was implemented the first time, e.g. blank lines in a message were removed, but if we set `replace_whitespace` to false, then this won't occur. However lines wrap in unexpected places once `replace_whitespace` is set to false (as you can see in the screenshot below), so this needs more investigation.

![Screenshot from 2020-04-23 11-57-49](https://user-images.githubusercontent.com/4522213/80138664-fbb8d780-8559-11ea-98c2-5fea8be7e4d7.png)

* Use `QPlainTextEdit` and its default wrapping feature that supports the kind of wrapping we want (a wrapping feature that never allows words to be cut off). See #1050 to understand the complications with updating the height of `QPlainTextEdit` to fit its contents. This also needs more investigation.

* Create our own wrapping method that allows us to wrap before cutting off a string if it doesn't already contain a wrapping point character. So basically we want to copy the automatic height resizing that `QLabel` supports and the automatic wrapping that never cuts off text that `QPlainTextEdit` supports.

As follow-up, @kushaldas and I will paste links to the Qt C++ code where all the relevant behavior takes place when it comes to wrapping and resizing.

**Comment by eloquence** [comment]
> Create our own wrapping method that allows us to wrap before cutting off a string if it doesn't already contain a wrapping point character. So basically we want to copy the automatic height resizing that QLabel supports and the automatic wrapping that never cuts off text that QPlainTextEdit supports.

Are there other cases than the one described in this issue (a long string without whitespace) that don't wrap correctly in  `SecureQLabel`? If it is just this issue, would it be sufficient to match for long strings without whitespace in the input, and insert a newline after a maximum number of characters that is safe to display at the minimum width of a bubble (e.g., 40)?

**Comment by creviera** [comment]
> * Use the library called textwrap to wrap message text every 70 characters. We saw problems with the way this was implemented the first time, e.g. blank lines in a message were removed, but if we set replace_whitespace to false, then this won't occur. However lines wrap in unexpected places once replace_whitespace is set to false

The[ textwrap docs](https://docs.python.org/3.1/library/textwrap.html#textwrap.TextWrapper.replace_whitespace) say: 

> If replace_whitespace is false, newlines may appear in the middle of a line and cause strange output. For this reason, text should be split into paragraphs (using str.splitlines() or similar) which are wrapped separately.

So this is the next step. However, it does look like there are other issues with using this library that we don't have when we rely on `QLabel` to wrap text. For instance, I grabbed all the test data in [this document ](https://github.com/minimaxir/big-list-of-naughty-strings/edit/master/blns.txt) (I even grabbed the comments because why not) sent it in a reply and saw this output:

![textwrap-every-70-chars](https://user-images.githubusercontent.com/4522213/80144508-7df9c980-8563-11ea-8fcd-d0f2e212148b.png)

Yet, when we use `QLabel` it looks correct (except for the one long-string wrapping issue described in the issue description above). `QLabel` seems to visually pass all our test-data tests, whereas `textwrap` doesn't. It is possible that following the `textwrap` doc instructions to split text into paragraphs (using str.splitlines() or similar) and calling `wrap` for each paragraph separately will fix this issue (however I doubt it'll be that easy). I'll continue to update the issue as I go as I make progress. 

Also, as you can see from the docs I linked to, `textwrap` is part of the Python standard library, which is very nice. 

> * Create our own wrapping method that allows us to wrap before cutting off a string if it doesn't already contain a wrapping point character. So basically we want to copy the automatic height resizing that QLabel supports and the automatic wrapping that never cuts off text that QPlainTextEdit supports.

The other solution where we create our own wrapping method will require more code that works around what I think is a bug in Qt (and apparently it was a [reported bug in Qt](https://bugreports.qt.io/browse/QTBUG-1630) which should be reviewed as a next step): `QLabel` should provide a way for us to use public methods to set the wrapping policy to match what is used by `QPlainTextEdit` so that strings are never cut off. To answer your question above, yes, there are wrapping points used by `QLabel` other than spaces, and it seems that these points vary based on characters that come before them, but as of now I've been able to see that text wraps on forward slashes `/`, hyphens `-`, and spaces ` `, so whichever one is nearest to the cutoff point is where `QLabel` will wrap. The next step is to review what happened around https://bugreports.qt.io/browse/QTBUG-1630. Then we should [find in the code](https://github.com/qt/qtbase/blob/60feaae1967aca5c7b1b0f6cb732962ddd1b2115/src/widgets/widgets/qlabel.cpp) where it decides to wrap on forward slash, dash, space, and possible other characters or sequence of characters (because I am unable to find this anywhere in the docs). We should also look into how `QPlainTextEdit` is able to wrap on a word boundary or at the cutoff point, see `WrapAtWordBoundaryOrAnywhere` [here](https://doc.qt.io/qt-5/qtextoption.html#WrapMode-enum) and [QTextDocument code](https://github.com/qt/qtbase/blob/60feaae1967aca5c7b1b0f6cb732962ddd1b2115/src/widgets/widgets/qplaintextedit.cpp#L761) as starting points.

> * Use QPlainTextEdit and its default wrapping feature that supports the kind of wrapping we want (a wrapping feature that never allows words to be cut off). See #1050 to understand the complications with updating the height of QPlainTextEdit to fit its contents. This also needs more investigation.

This solution requires resizing `QPlainTextEdit` to fit its contents. At first look, this seems simpler than using `QLabel` since wrapping is already done the way we want it to be done. All we need to do is get the height of the document after it has wrapped its contents to fit within the width of the widget and set the height of `QPlainTextEdit` to equal this. The challenge here is making sure we get a correct pixel width of special characters such as a Tab (instead of getting the pixel width of  `U+0009` we want the pixel width of the space between the first character before the tab and the next tab stop. @redshiftzero shared in #1050 a way to use `QFontMetrics.size` in order to get the pixel width of the tab once it's been expanded, but we'll need to look into doing this for other special characters.

--

So that's the latest for each investigation.

**Comment by eloquence** [comment]
Thanks for the excellent summary @creviera :). Yeah, the fact that other characters wrap long lines in QLabel surely complicates the approach of rolling our own logic.

In my own testing I can confirm that long URLs do in fact typically wrap well right now, which is good news for the pilot and somewhat reduces the severity of this issue in practice.

I agree it's worth continuing to investigate both approaches, especially if we feel that QPlainTextEdit offers us other benefits. But we may want to time-box that into chunks, and consider shipping a release of the Client without a full resolution to this issue.

**Comment by creviera** [comment]
For now, I think we could document this as a known issue with some examples of what you might see when there is a long string such as: `https://something.com/gBbRmlsZSBzb3VyY2U6IGh0dHBzOi8vY29tbW9ucy53aWtpbWVkaWEub3JnL3dpa2kvRmlsZTpMb3dfcHJlc3N1cmVfc3lzdGVtX292ZXJfSWNlbGFuZC5qcGf/etc` which will look like:

![Screenshot from 2020-04-23 15-21-36](https://user-images.githubusercontent.com/4522213/80155243-2e70c900-8576-11ea-8885-379a9c36269f.png)

The way to get the full string as of now in the latest nightly is to copy all the text in the message and paste it somewhere (and evaluate the string before visiting the link). Also, this example of a URL isn't your average URL. Most will not have strings as long as `gBbRmlsZSBzb3VyY2U6IGh0dHBzOi8vY29tbW9ucy53aWtpbWVkaWEub3JnL3dpa2kvRmlsZTpMb3dfcHJlc3N1cmVfc3lzdGVtX292ZXJfSWNlbGFuZC5qcGf` and will wrap on `/`s so that you can see the full URL.

**Comment by creviera** [comment]
Also, if anyone wants to contribute to our collection of test data, which will use for unit tests and our test data for staging servers, please add to it here: 

https://github.com/freedomofpress/securedrop-client/wiki/Test-plan#input-testing

**Comment by creviera** [comment]
Following up on using `textwrap` (I don't recommend that we use this library and I explain why below):

You can see on this branch https://github.com/freedomofpress/securedrop-client/tree/use-textwrap-lib that many of input tests (https://github.com/freedomofpress/securedrop-client/wiki/Test-plan#input-testing) pass, but there are failures around a single issue: `textwrap` wraps text at a hard-set character count rather than a pixel width. In order to get this right, we would either have to wrap at somewhere around 40 characters (which looks bad in most cases) or we would have to iterate through each character in the input text and calculate its width and only add a newline when there it exceeds the width of the QLabel. At that point, there's no reason to use `textwrap`.

So this string passes the test:

```
thisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashesthisisalongwordwithoutspacesordashes
```

But this string does not:

```
THISISALONGWORDWITHWIDECHARACTERSTHISISALONGWORDWITHWIDECHARACTERSTHISISALONGWORDWITHWIDECHARACTERSTHISISALONGWORDWITHWIDECHARACTERSTHISISALONGWORDWITHWIDECHARACTERSTHISISALONGWORDWITHWIDECHARACTERSTHISISALONGWORDWITHWIDECHARACTERSTHISISALONGWORDWITHWIDECHARACTERS
```

Also, when you copy the string above from a speech bubble, you'll see newlines in incorrect places: 

```
THISISALONGWORDWITHWIDECHARACTERSTHISISALONGWORDWITHWIDECHARACTERSTHIS
ISALONGWORDWITHWIDECHARACTERSTHISISALONGWORDWITHWIDECHARACTERSTHISISAL
ONGWORDWITHWIDECHARACTERSTHISISALONGWORDWITHWIDECHARACTERSTHISISALONGW
ORDWITHWIDECHARACTERSTHISISALONGWORDWITHWIDECHARACTERS
```

**Comment by creviera** [comment]
So, actually, there's really two issues with using `textwrap`: (1) we want to wrap when the size of the text reaches the maximum _pixel-width_, not character-width, and (2) the addition of newlines changes the text that will be copied to a user's clipboard.

If we use Qt to wrap text we don't have these issues.

**Comment by redshiftzero** [comment]
Hmm nice investigation. So it seems like the `textwrap` library is really only useful for terminal environments where we're using fixed width fonts. Agreed that it doesn't really fit our needs. 

> we would have to iterate through each character in the input text and calculate its width and only add a newline when there it exceeds the width of the QLabel

This seems to me to now be to be the most promising option: we know how to do this from the elided text functionality (and from our funvestigations in #1050), and if we use a subclass of `QLabel` then we don't need to worry about explicitly setting the height of the bubble as we did with `QPlaintextEdit` or using the more complex functionality in the web view in #1050 (complex just because of the Qt widget in use, not the implementation in #1050 :heart:). What do you think @creviera @kushaldas, does that seem like a good path?

**Comment by ntoll** [comment]
Excuse the interruption. I'm somehow still getting emails linked to this issue.... (will reconfigure my settings) ...but cheering you on from afar. You folks rock! Best of luck! :snake: :heart: :rocket: :+1:

**Comment by creviera** [comment]
@redshiftzero - I think we should continue to use `QLabel` because the issue described here in #815 is more of an edge case, and resizing `QPlainTextEdit` or switching to `QWebEngineView` will take some time to get right.

In order to solve the `QLabel` wrapping issue (note: I think the solution described in this bug report is the proper way to fix it: https://bugreports.qt.io/browse/QTBUG-1630) we would have to do some prepocessing of the input text in order to see whether or not a word exists that is longer than the width of the `QLabel` and has no wrapping points (wrapping points are forward slash, space, dash, and special characters such as `$`), then we can add our own newlines in the middle of those words where needed. Unfortunately, this does introduce issue 2 described above: https://github.com/freedomofpress/securedrop-client/issues/815#issuecomment-620879618. In order to get around this issue, we would have to use a class derived from `QTextDocument` such as `QPlainTextEdit` which has a `WrapAtWordBoundaryOrAnywhere` setting that properly wraps text without adding newlines. 

At this point, I think it would be better to wait on moving forward with a workaround until we file a bug report with Qt. I think this should be reopened: https://bugreports.qt.io/browse/QTBUG-1630, and I can describe how switching to a QTextDocument doesn't provide a way to resize the widget to fit the text. I also think the `QWebEngineView` has a lot of promise, but I agree that we would need to take time to lock it down since it provides more complex functionality. I also found an issue with it not wrapping on words that use characters only available in fonts that are missing on your system. Instead of wrapping the blank rectangles that are used in place of missing characters, it created a horizontal scroll on the message bubble. @kushaldas could provide a better estimate on how long it would take to fix this issue, but I think for now we should write #815 up as a know issue, file a bug with Qt, reach out to the Qt community, and wait on creating a workaround.

**Comment by redshiftzero** [comment]
> I think this should be reopened: https://bugreports.qt.io/browse/QTBUG-1630, and I can describe how switching to a QTextDocument doesn't provide a way to resize the widget to fit the text.

Fair points, let's do that. Want to file that upstream? Until then we can put this issue on hold, it's worth noting that the priority of this is decreased a bit since now with copy/paste support the user _can_ at least copy that text.

**Comment by eloquence** [comment]
We've agreed to add upstream follow-up (not resolution) to the 5/6-5/20 sprint.

**Comment by eloquence** [comment]
I'll be working with @creviera to do an upstream write-up.

**Comment by eloquence** [comment]
With help from @creviera I've posted a note upstream, here:
https://bugreports.qt.io/browse/QTBUG-1630

That's our only commitment in this sprint, so I'm moving this issue back to the backlog for the time being, and we can decide what to do if/when we get a response from upstream.

**Comment by eloquence** [comment]
Cross-ref additional Qt bug filed by @creviera: https://bugreports.qt.io/browse/QTBUG-85498

Nothing much to do here but wait a bit, so moving back to near-term for now.

**Comment by eloquence** [comment]
Per quick chat with @creviera, based on upstream feedback https://bugreports.qt.io/browse/QTBUG-85498?focusedCommentId=520391&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-520391, it may make sense to do an additional brief implementation spike to implement custom wrapping logic for special cases Qt doesn't handle well, nominating for discussion tomorrow at sprint planning.

**Comment by eloquence** [comment]
^ Per the above, @creviera has committed to allocating a 2 hour timebox during the 7/23-8/5 sprint to see if this approach is viable.

**Comment by creviera** [comment]
Just pointing out that this workaround is similar to https://github.com/freedomofpress/securedrop-client/pull/838, but instead of adding newlines every 70 characters we can just add newlines _only_ on strings that do not contain a space (later we can make sure to avoid inserting a new line on strings that will wrap because they have other wrapping point characters such as a dash, a slash, etc. - a space is not the only wrapping point) and are more than a specified length. We want to avoid adding newlines if we don't have to because it alters the text that a user might copy and paste somewhere else. Another issue with this workaround is that we will have to insert newlines in the middle of a line in the speech bubble in order to account for speech bubble resizing behavior that we will introduce in order to better support smaller screen resolutions. This will not look as nice as wrapping once the line of text reaches the length of the speech bubble. 

A workaround for the first issue mentioned above for copy and paste would be to create a copy function that grabs the original, unaltered text rather than the text from the speech bubble. A workaround for the second issue to support responsive speech bubbles would be to listen for a resize event and to re-calculate the cut-off point based on the new width - this is not great, so we'll probably just stick to adding newlines wherever our minimum speech bubble width is set.

The underlying fix for this issue is being tracked here: https://bugreports.qt.io/browse/QTBUG-85498.

**Comment by eloquence** [comment]
> (later we can make sure to avoid inserting a new line on strings that will wrap because they have other wrapping point characters such as a dash, a slash, etc. - a space is not the only wrapping point) 

I agree it would be a nice iteration to not wrap in cases where we can detect that Qt's built-in behavior will do a better job than our insertion of a newline at the `n` character mark. 

> we'll probably just stick to adding newlines wherever our minimum speech bubble width is set.

That makes sense to me. I don't think we need to ever support a lower window size than 1280x1024. It would be good to know what line length we can render at that resolution (given maximum character width), and whether or not that's a reasonable wrapping point.

I think the priority of changing the copy/paste behavior would depend on how likely a user is to encounter this behavior in the real world. If it's extremely unlikely except for highly unusual URLs or strings, I don't think we need to prioritize changing the copy/paste behavior to copy the raw (unwrapped) original.

**Comment by creviera** [comment]
> That makes sense to me. I don't think we need to ever support a lower window size than 1280x1024. It would be good to know what line length we can render at that resolution (given maximum character width), and whether or not that's a reasonable wrapping point.

We did end up adding support for a 1366 x 768 resolution, see https://github.com/freedomofpress/securedrop-client/issues/1074#issue-607732712, but as of now you have to horizontally scroll in order to see the full width of the messages and replies. If we want to support showing all of the text in the speech bubbles, then we'll either have to change the speech bubble width to make it more responsive, or change the size of the sourcelist, or change the size/hide the left bar, or leave the horizontal scrolling as it is now. Hopefully by the time we get to this the Qt bug will be resolved. For now, we can just assume the width of the speech bubbles are static.

**Comment by creviera** [comment]
here is a spike i did of speech bubbles that change in size based on size of window, for example:

![responsive-client](https://user-images.githubusercontent.com/4522213/88585182-9c121280-d007-11ea-8ed8-60ca6a015ea8.png)

**Comment by creviera** [comment]
i'm going to stop the spike at this point so we can check in tomorrow:

* as we found in https://github.com/freedomofpress/securedrop-client/issues/1074 not all characters are the same width so some long words will still get cut off. to get around this, we can use a very short max line length (in characters) that would make it less likely to happen.
* we can get word length by looking at `fm.size(Qt.TextSingleLine, word).width()` but then we need to figure out where to add the newline, so what we actually want to do is parse each character in the text, which is simiilar to the workaround offered in the bug report, but it doesn't seem to take into account non-word characters and it's much less sophisticated than Qt's current implementation for wrapping
* i still think we should stick to the plan to document this as a known upstream bug and wait for our bug report to be addressed while we work on larger securedrop issues

**Comment by eloquence** [comment]
Thanks much for the spike @creviera and the upstream follow-up. We checked in on this issue with @emkll today and agreed to again park this again on the backlog for now, giving upstream additional time to respond.

To add to Allie's recap: The workaround to preempt wrapping certain long strings before Qt does it risks introducing regressions for some long strings that are currently wrapped nicely (e.g., very long URLs wrapped at a `/`). In those cases we'd go from a clean wrap without a newline, to a dirty (potentially mid-word) wrap with a newline. It also seems rather finicky to match custom wrapping against Qt's.

We don't even fully understand under what circumstances we'd be overriding Qt's behavior, since we still don't have a full list of Qt's wrap point characters.

A simpler workaround may be to just enable horizontal scrollbars for a message bubble *iff* (if and only if) a message includes an overlong, unwrapped string. We would want to ensure that other strings that are part of the same message are still wrapped correctly, without a need for scrolling. This is not a very high priority issue but one we'll ultimately want to resolve, so we'll check in again in the n+1 or n+2 sprint.

**Comment by eloquence** [comment]
I'm not sure if this is a regression or not, but when following the test plan for https://github.com/freedomofpress/securedrop-workstation/pull/751 on `securedrop-client_0.6.0-dev-20220331-060621`, I observed that when sending a reply like 

`0123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789`

I cannot scroll to reveal its whole contents. I can use Ctrl+A/Ctrl+C to select the whole content and then copy and paste it into the app to see it, though. Is that the currently expected behavior, or was it previously possible to scroll within these replies to see the contents?

**Comment by creviera** [comment]
Yup this is exactly the issue being tracked upstream in https://bugreports.qt.io/browse/QTBUG-85498 (I updated the title to make that more clear). If I recall correctly, the workarounds are to either switch to `QPlainTextEdit` which will wrap long strings of alphanumerics that have no spaces or to create or own `QLabel` wrapping function that adds newlines in appropriate locations (which we would want to remove when a user copies the content of the message).

**Comment by eloquence** [comment]
OK cool. I'll update our test plan to make it clearer that long, uninterrupted strings will be cut off.

**Comment by eloquence** [comment]
Closing, non-critical issue specific to the current codebase/UX, which is undergoing a rewrite.
